plugins/api/src/switch_user.cpp

```diff
@@ -0,0 +1,256 @@
+            auto& client = _comm->clientUser;
+
+            // Update the user identity associated with the RsComm.
+            std::strncpy(client.userName, _input->username, sizeof(UserInfo::userName));
+            std::strncpy(client.rodsZone, _input->zone, sizeof(UserInfo::rodsZone));
+            std::strncpy(client.userType, user_type_string, sizeof(UserInfo::userType));
+
+            // Set the appropriate privilege level based on whether the user is local or remote
+            // to this zone.
+            if (std::strncmp(_input->zone, local_zone, sizeof(SwitchUserInput::zone)) == 0) {
+                client.authInfo.authFlag =
+                    (*user_type == adm::user_type::rodsadmin) ? LOCAL_PRIV_USER_AUTH : LOCAL_USER_AUTH;
+            }
+            else {
+                client.authInfo.authFlag =
+                    (*user_type == adm::user_type::rodsadmin) ? REMOTE_PRIV_USER_AUTH : REMOTE_USER_AUTH;
+            }
+
+            // iRODS agents do not disconnect from other nodes following a redirect. For long running agents,
+            // this means the API plugin must invoke rc_switch_user() on each connection.
+            update_server_to_server_connections(*_comm);
+
+            return 0;
+        }
+        catch (const irods::exception& e) {
+            log_api::error(e.what());
+            return e.code();
+        }
+        catch (const std::exception& e) {
+            log_api::error(e.what());
+            return SYS_LIBRARY_ERROR;
+        }
+        catch (...) {
+            log_api::error("An unknown error occurred while processing the request.");
+            return SYS_UNKNOWN_ERROR;
+        }
+    } // rs_switch_user
+
+    using operation = std::function<int(RsComm*, SwitchUserInput*)>;
+    const operation op = rs_switch_user;
+
+// clang-format off
+#define CALL_SWITCH_USER call_switch_user
+// clang-format on
+} // anonymous namespace
+
+#else // RODS_SERVER
+
+//
+// Client-side Implementation
+//
+
+namespace
+{
+    using operation = std::function<int(RsComm*, SwitchUserInput*)>;
+    const operation op{};
+
+// clang-format off
+#define CALL_SWITCH_USER nullptr // NOLINT(cppcoreguidelines-macro-usage)
+// clang-format on
+} // anonymous namespace
+
+#endif // RODS_SERVER
+
+// The plugin factory function must always be defined.
+extern "C" auto plugin_factory(
+    const std::string& _instance_name, // NOLINT(bugprone-easily-swappable-parameters)
+    const std::string& _context) -> irods::api_entry*
+{
+    static_cast<void>(_instance_name);
+    static_cast<void>(_context);
+
+#ifdef RODS_SERVER
+    irods::client_api_allowlist::instance().add(SWITCH_USER_APN);
+#endif // RODS_SERVER
+
+    // clang-format off
+    irods::apidef_t def{SWITCH_USER_APN,            // API number
+                        RODS_API_VERSION,           // API version
+                        NO_USER_AUTH,               // Client auth
+                        NO_USER_AUTH,               // Proxy auth
```

Comment:
Consider these.

Right now, privilege checks for handled by the API plugin directly. However, setting these allows us to remove that check.

The current implementation is easier to reason about. Hmm ...

Reply:
The error message on 129 is very good - will explain clearly when someone is holding it wrong...   is there a way to get that level of explanation here?

I think it's probably fine the way it is now if you leave a C++ comment near 236/237 saying what you've got in the discussion comment in this thread - "the 'real' check is above"

Reply:
Newer code has used the style of L129 because it's easy/flexible, but it side steps the design.

With L236/237, we can't report error messages like with L129. However, we could modify the framework and apidef_t type to hold **static** error messages.

I'm fine with L129 though, it works. I just remembered the other way of doing this and it made me wonder.

I'm going to stick with L129.

Reply:
Seems good.